openmls/src/binary_tree/array_representation/diff.rs

```diff
@@ -0,0 +1,634 @@
+
+        NodeReference::try_from_node_index(self, copath_node_index)
+    }
+
+    /// Returns a vector of [`NodeReference`]s, where the first reference is to
+    /// the root of the shared subtree of the two given leaf indices followed by
+    /// references to the nodes in the direct path of said subtree root.
+    ///
+    /// Returns an error if either of the two given leaf indices do not
+    /// correspond to a leaf in the diff.
+    pub(crate) fn subtree_path(
+        &self,
+        leaf_index_1: LeafIndex,
+        leaf_index_2: LeafIndex,
+    ) -> Result<Vec<NodeReference>, ABinaryTreeDiffError> {
+        let node_index_1 = to_node_index(leaf_index_1);
+        let node_index_2 = to_node_index(leaf_index_2);
+
+        self.out_of_bounds(node_index_1)?;
+        self.out_of_bounds(node_index_2)?;
+
+        let lca = lowest_common_ancestor(node_index_1, node_index_2);
+        let direct_path_indices = direct_path(lca, self.size())?;
+        let mut full_path = vec![NodeReference::try_from_node_index(self, lca)?];
+        for node_index in direct_path_indices {
+            let node_ref = NodeReference::try_from_node_index(self, node_index)?;
+            full_path.push(node_ref);
+        }
+
+        Ok(full_path)
+    }
+
+    // Functions pertaining to the whole diff
+    /////////////////////////////////////////
+
+    /// Returns an iterator over references to the content of all nodes in the
+    /// diff.
+    pub(crate) fn iter(&'a self) -> DiffIterator<'a, T> {
+        DiffIterator {
+            diff: self,
+            current_index: 0u32,
+        }
+    }
+
+    /// Returns a vector containing the nodes of the tree in-order, i.e. in the
+    /// array representation of the diff. This function should not fail and only
+    /// returns a [`Result`], because it might throw a
+    /// [LibraryError](ABinaryTreeError::LibraryError).
+    pub(crate) fn export_nodes(&self) -> Result<Vec<T>, ABinaryTreeDiffError> {
+        let mut nodes = Vec::new();
+        for node_index in 0..self.size() {
+            let node = self.node_by_index(node_index)?;
+            nodes.push(node.clone());
+        }
+        Ok(nodes)
+    }
+
+    /// Returns the size of the diff.
+    pub(in crate::binary_tree) fn size(&self) -> NodeIndex {
+        self.size
+    }
+
+    /// Returns the leaf count of the diff.
+    pub(crate) fn leaf_count(&self) -> LeafIndex {
+        (self.size() + 1) / 2
+    }
+
+    // Functions around individual [`NodeReference`]s
+    ///////////////////////////////////////////////
+
+    /// Returns a reference to the node pointed to by the [`NodeReference`].
+    /// Returns an Error if the [`NodeReference`] points to a node outside of the
+    /// bounds of the tree. This can happen, for example, if the node was
+    /// removed while shrinking the diff after the creation of the
+    /// [`NodeReference`].
+    pub(crate) fn try_deref(&self, node_ref: NodeReference) -> Result<&T, ABinaryTreeDiffError> {
```

Comment:
Would it make sense to implement the `Deref`/`DerefMut` traits here instead?

Reply:
Ok I see. Yes you'll need keep the `&self` in scope. I think it's the function name again 😉 
How about just `node`?